### ISSUE TITLE
Hide `if err != nil { return err}` like goland

### ISSUE THREAD

**Issue by guettli** [golang/vscode-go#2311]
I am evaluating vscode and goland at the moment. This feature of goland is nice:

![image](https://user-images.githubusercontent.com/414336/176840822-ad11e330-a654-415b-99c7-5889cba8a472.png)

Only if I expand the block, I see

![image](https://user-images.githubusercontent.com/414336/176841019-65b78ba9-3a73-42b0-9256-a63597a5d69b.png)

This feature of goland reduces the amout of characters which my eyes need to parse. If you would sum all these
characters which my eyes don't need to parse per day, this would be a lot. A lot less cognitive load.

It would be great, if vscode would support this feature, too.

**Comment by hyangah** [comment]
@guettli Thanks for the feedback! Can you help us understand exactly what aspect of the folding feature in goland you like?
  * Folded by default?
  * Unlike VSCode Go that leaves the closing `}` when folded, goland completely hides `{}`?
  * Or presenting helpful collapsed text e.g. ": err ..." when folded?
    
![Screen Shot 2022-07-06 at 2 09 46 PM](https://user-images.githubusercontent.com/4999471/177615446-5fedca6c-3c5d-4401-84e8-130aa4f66134.png)

(VSCodeGo/gopls currently doesn't utilize the `collapsedText` feature yet and shows "..." cc @golang/tools-team )

**Comment by taqqanori** [comment]
l'm just thinking the same, and created an extension, although it uses built-in `editor.fold` command (leaves `}` unfolded).
https://marketplace.visualstudio.com/items?itemName=taqanori.hide-error-cases

**Comment by jogly** [comment]
> Folded by default?
Unlike VSCode Go that leaves the closing } when folded, goland completely hides {}?
Or presenting helpful collapsed text e.g. ": err ..." when folded?

Hopefully resurrecting this instead of making a new feature request 🙏 .  @hyangah - your first two points are what would be most useful to me.  The third would be nice to have. The second point is the most painful of the three; the dangling brace makes folding three lines into two not worth the effort. FWIW, I don't care if the braces are hidden or not; it's the trailing brace on it's own line that hurts readability. For example, the following are equivalently useful to me:
```
if err != nil {...}
```
```
if err != nil : err ⤴︎
```

**Comment by suzmue** [comment]
The LSP spec also now has (proposed?) support for collapsedText in folding range results as of 3.17.0 https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_foldingRange

**Comment by guettli** [comment]
I think foldingRange of LSP should be supported now by vscode.

It would be great, if you could fold/unfold lines which return errors.

**Comment by firelizzard18** [comment]
I'd like to take a look at implementing this, though I'm not sure when I'll have time (I'm a community contributor despite the "member" tag).

**Comment by sebastian-philipp** [comment]
Why not also hide the `if err != nil {...}`part and reduce it to: `err ⤴︎`


like from: 
```go
data, ok := something.Data["data"]
if !ok {
	return nil, errors.New("missing data")
}

cfg, err := GetCfg(data)
if err != nil {
	return nil, errors.Wrap(err, "failed to get cfg")
}

scheme := runtime.NewScheme()
err = clientgoscheme.AddToScheme(scheme)
if err != nil {
	return nil, errors.Wrap(err, "failed to add client-go scheme")
}
err = other.AddToScheme(scheme)
if err != nil {
	return nil, errors.Wrap(err, "failed to add other scheme")
}

client, err := something.New(cfg, scheme)
if err != nil {
	return nil, errors.Wrap(err, "failed to create client")
}
```
to: 

```go
data, ok := something.Data["data"]
if !ok {
	return nil, errors.New("missing data")
}

cfg, err := GetCfg(data)
err ⤴︎

scheme := runtime.NewScheme()
err = clientgoscheme.AddToScheme(scheme)
err ⤴︎
err = other.AddToScheme(scheme)
err ⤴︎
client, err := something.New(cfg, scheme)
err ⤴︎
```

**Comment by firelizzard18** [comment]
Personally I find that a lot less readable. I could possibly see folding it into the previous line but folding it like you suggest makes it a lot less obvious to me at a glance what's happening.

**Comment by h9jiang** [comment]
Hi, I did some research based on the discussion above and want to experiment with how this can be implemented. 

Current behavior, the code snippet below will be folded as two lines. 

```
// before folded
if err := function(); err != nil {
   return err
}

// folded
if err := function(); err != nil {...
}
```


Based on [comment](https://github.com/golang/vscode-go/issues/2311#issuecomment-1368022844) desired behaviors are: 
1. when folded, the `}` should not occupied it's own line. The code snippet above should be folded as `if err := function(); err != nil {...}` in one line instead of two lines.
2. error handling should be folded by default. 
3. nice to have: provide helpful collapsed text.

Here are some limitations I found from VSCode.
1. `collapsedText` is now supported by  LSP [FoldingRangeClientCapabilities](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#foldingRangeClientCapabilities), but not supported by VSCode. FR in microsoft/vscode-languageserver-node#1068 microsoft/vscode#70794. 
2.`lineFoldingOnly` being marked as true in [vscode-languageserver-node](https://github.com/microsoft/vscode-languageserver-node/blob/d810d51297c667bd3a3f46912eb849055beb8b6b/client/src/common/foldingRange.ts#L40) meaning the gopls will only send line info instead of further detailed starting character and ending character. 
3. The current `foldingRangeKind` set is very limited. We want to introduce more kind to help us understand the code better. 

Let's come back to the desired behavior:
1. Gopls can return an extra line if the closing `}` is the only character in that line. Right now, gopls is returning `{start: 0, end: 1, kind: unknown}`, if gopls can return `{start: 0, end: 2, kind: unknown}`. The VSCode will be able to fold the code in form of 

```
if err := fmt.Errorf(""); err != nil {...
```

This is kind of annoying because I won't be able to see the `}`. If the `collapsedText` is implemented, we can customize the text using `...}` (Unfortunately the `collapsedText` it's not implemented yet). 

2. Gopls can return extra folding kind information instead of unknown. Like `{start: 0, end: 1, kind: error-handling}` or simply `default-folded`. So vscode-go can call `editor.Fold` on folding range of this type. vscode-go can also trigger this `editor.Fold` command when a file get opened. So by default, error handling folding range will be folded. 

3. Some more thinking need to do in Gopls side when determine what should be considered as a range should be folded by default. Like, gopls should consider the statements inside of error handling block (error handling with one return statement can be folded by default), gopls should consider the function signature (error handling with one return statement and every other return value except for error is zero value can be folded by default) ....

I'm new to the VS Code API, so please feel free to correct me or suggest alternative solutions if I'm off track.

**Comment by firelizzard18** [comment]
The heuristic I would choose is:

- The function's last return value is `error`.
- `(*IfStmt).Body` contains a single statement and that statement is a `*ReturnStmt`.
- The last `(*ReturnStmt).Results` is a value/something other than `nil`.
- Consider a followup CL to add support for variants such as `{ err = fmt.Errorf(...); return err }`.
- It would be nice to validate that the error return value is non-nil (e.g. excluding `if err == nil { return err }`) but maybe that can be another followup.

Personally I wouldn't worry about the other return values. IMO the error conditional in the following example should be folded. In summary: "If a conditional checks err != nil and returns an error, it is considered to be an error handling statement regardless of other return values." My logic is, there are circumstances where it makes sense to return non-zero values for other parameters from an error handling statement, so gopls should respect that.

```golang
func Foo(x int) (int, error) {
	err := Bar(x)
	if err != nil {
		return -1, err
	}
	return x, nil
}
```

**Comment by eihigh** [comment]
Since the Go team has decided not to improve the language's error handling syntax, editor support for this has become even more important. Is anyone here still interested in this issue?
https://go-review.googlesource.com/c/website/+/673615

**Comment by adonovan** [comment]
This seems reasonable as an optional feature. If someone sends a CL to use FoldingRange's CollapsedText feature, I'll review it.

**Comment by firelizzard18** [comment]
I previously said I was interested in working on this but unfortunately I don't think I'll have the bandwidth for a while.

**Comment by myaaaaaaaaa** [comment]
An alternative approach would be to lower the opacity of `err != nil` blocks, except for when the mouse hovers over them or when the cursor is inside them.

See the below demonstration for details:

https://github.com/user-attachments/assets/c1351b72-544a-4405-8fea-23ea81ed229f

**Comment by ufukty** [comment]
@myaaaaaaaaa You might consider that on-hover behaviour because of performance concerns. From my previous experience on developing a dimming extension for vscode, I can tell that the performance impact might be significant enough to annoy user, or cause user to lose focus on rare but important occasions. It might land below the line where the value meets cost. Other hand, skipping only the focused/selected areas provides good balance on that ratio.

**Comment by adonovan** [comment]
Another possibility would be for the server's semantic tokens query to return an optional modifier for all the tokens in `if err != nil { ... }` that would allow the client to render it differently somehow. Might be worth a quick experiment.

**Comment by firelizzard18** [comment]
I think it's valuable to discuss improving folding (this issue) and purely visual/color/opacity changes (#3793) separately, since the implementation of those is different and as I stated in the other issue I can see use cases for implementing both.

**Comment by kaixuan188** [comment]
Is there any new progress?

**Comment by adonovan** [comment]
> Is there any new progress?

No, there is not; sorry.

**Comment by firelizzard18** [comment]
Without https://github.com/microsoft/vscode/issues/70794 or https://github.com/microsoft/vscode/issues/3352 being addressed, the best the extension can achieve is:

```go
if err := fmt.Errorf(""); err != nil {...
```

(It could fold everything into a single line.) VSCode is only capable of folding whole lines, not character ranges, and it does not support custom replacement text, so that snippet above is the best the extension could do. Until one or both of those vscode issues is resolved, there's not much we can do. **TL;DR: Status: waiting on upstream.**

**Comment by mohsinkaleem** [comment]
is there any update on this error-block folding feature?
would be of great help for many of us

**Comment by adonovan** [comment]
> is there any update on this error-block folding feature? would be of great help for many of us

As the previous comment explained, this is blocked on upstream issues in VS Code. If you want to apply pressure, those issues are the place to do it.
